plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/SizeBasedBlockingQ.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.repositories.s3.async;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.core.common.unit.ByteSizeValue;
+import org.opensearch.repositories.s3.GenericStatsMetricPublisher;
+import org.opensearch.repositories.s3.S3TransferRejectedException;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Queue implementation to accept events based on their storage attribute. If size of queue is breached, then transfer
+ * event is rejected.
+ */
+public class SizeBasedBlockingQ extends AbstractLifecycleComponent {
+    private static final Logger log = LogManager.getLogger(SizeBasedBlockingQ.class);
+
+    protected final LinkedBlockingQueue<Item> queue;
+    protected final Lock lock;
+    protected final Condition notEmpty;
+
+    protected final AtomicLong currentSize;
+    protected final ByteSizeValue capacity;
+    protected final AtomicBoolean closed;
+    protected final ExecutorService executorService;
+    protected final int consumers;
+    private final GenericStatsMetricPublisher genericStatsMetricPublisher;
+    private final Type type;
+
+    /**
+     * Constructor to create sized based blocking queue.
+     */
+    public SizeBasedBlockingQ(
+        ByteSizeValue capacity,
+        ExecutorService executorService,
+        int consumers,
+        GenericStatsMetricPublisher genericStatsMetricPublisher,
+        Type type
+    ) {
+        this.queue = new LinkedBlockingQueue<>();
+        this.lock = new ReentrantLock();
+        this.notEmpty = lock.newCondition();
+        this.currentSize = new AtomicLong();
+        this.capacity = capacity;
+        this.closed = new AtomicBoolean();
+        this.executorService = executorService;
+        this.consumers = consumers;
+        this.genericStatsMetricPublisher = genericStatsMetricPublisher;
+        this.type = type;
+    }
+
+    public enum Type {
+        NORMAL,
+        LOW;
+    }
+
+    @Override
+    protected void doStart() {
+        for (int worker = 0; worker < consumers; worker++) {
+            Thread consumer = new Consumer(queue, currentSize, lock, notEmpty, closed, genericStatsMetricPublisher, type);
+            executorService.submit(consumer);
+        }
+    }
```

Comment:
Should we use a scaling threadpool instead?

Reply:
To effectively use scaling we have to think of creating consumers dynamically. Currently, these are constant and created on node start.and that's why I used fixed